File: src/ffi/alerts/peer_disconnected.rs
#[cxx::bridge(namespace = "ltrs")]
pub(crate) mod ffi {
    #[namespace = "libtorrent"]
    extern "C++" {
        type peer_disconnected_alert = crate::ffi::ffi::peer_disconnected_alert;
    }

    unsafe extern "C++" {
        include!("cpp/alerts/peer_disconnected.h");
        type Error = crate::ffi::error::ffi::Error;

        unsafe fn peer_disconnected_alert_get_socket_type(
            alert: *mut peer_disconnected_alert,
        ) -> u8;
        unsafe fn peer_disconnected_alert_get_op(alert: *mut peer_disconnected_alert) -> u8;
        unsafe fn peer_disconnected_alert_get_error(alert: *mut peer_disconnected_alert) -> Error;
        unsafe fn peer_disconnected_alert_get_reason(alert: *mut peer_disconnected_alert) -> u16;
    }
}

File: cpp/alerts/peer_disconnected.cpp
#include "./peer_disconnected.h"
#include "cpp/error.h"
#include <cstdint>

namespace ltrs {

uint8_t peer_disconnected_alert_get_socket_type(lt::peer_disconnected_alert* alert) {
    return static_cast<uint8_t>(alert->socket_type);
}

uint8_t peer_disconnected_alert_get_op(lt::peer_disconnected_alert* alert) {
    return static_cast<uint8_t>(alert->op);
}

Error peer_disconnected_alert_get_error(lt::peer_disconnected_alert* alert) {
    return error_code_to_error(alert->error);
}

uint16_t peer_disconnected_alert_get_reason(lt::peer_disconnected_alert* alert) {
    return static_cast<uint16_t>(alert->reason);
}

} // namespace ltrs
